block-layer/codecs/dag-pb.md

```diff
@@ -44,13 +44,13 @@ This layout can be expressed with [IPLD Schemas](../../schemas/README.md) as:
 
 ```ipldsch
 type PBLink struct {
-  Hash Bytes
-  Name optional String
-  Tsize optional Int
+  Hash optional Link
+  Name String (implicit "")
+  Tsize Int (implicit "0")
 }
 
 type PBNode struct {
-  Links optional [PBLink]
+  Links [PBLink]
   Data optional Bytes
```

Comment:
https://github.com/ipfs/go-merkledag/blob/749fd8717d46b4f34c9ce08253070079c89bc56d/pb/merkledag.pb.go#L154-L159

Reply:
Is it possible to have a zero-length `Data` element in go-ipld-prime-proto? So zero-length and empty/nil are both valid forms? In JS zero-length is currently not allowed (it's converted to a `null`).

Reply:
The type in Go is `[]byte`, which indeed can be either `nil` or `[]byte{}` (empty and non-nil list), but I don't think the Protobuf encoding separates those two. From the proto3 language guide:

> For bytes, the default value is empty bytes.

So I think whichever you choose will always be encoded as empty bytes in protobuf, I imagine. But I'm not sure.

Reply:
I also get the impression from [the protobuf language guide](https://developers.google.com/protocol-buffers/docs/proto#optional) that the wire format will not disambiguate between an absent value here and an empty byte sequence.

I don't emotionally _like_ this, because it would feel a lot "cleaner" if our IPLD'ified views could give a clear `Absent` value, but I think the [same logic I just used about empty lists](https://github.com/ipld/specs/pull/297/files#r494338322) applies here as well.

Reply:
Not quite, `0x0a00` means - "field 1, wire type 2 (bytes), length 0": `{Data:[]bytes{}}`

Where `0x1200` means - "field 2, wire type 2 (bytes), length 0", but for field 2 it has to drop into parsing the `Links` array with a zero byte array and that yields a single empty link: `{Links:[{}]}`

So in terms of cardinality you can't avoid the absent, zero-length, non-zero-length cases for `Data`.

For `Links` you can have absent but you can't have zero-length.